File: Python Compiler/Compiler.py
# TODO: add error handling

import os
import array

# Ensure you set the current directory to where your file is located
os.chdir("../")

# Reads the target file to compile
with open("target.bf", "r") as f:
    code = f.read()

# Converts the file to a string and removes all non valid characters (thus 'free' code)
code_free = [i for i in code if i in "><+-.,[]"]

# Prints the file (without comments) (for testing purposes)
print("".join(code_free), "\n")

# Initializes the memory
memory = array.array("B", [0] * 30000)  # Extendable memory (change if more memory is needed)
# Initializes the memory pointer
cur_pointer = 0

# Preprocess the code to find matching brackets
bracket_map = {}
stack = []
for i, command in enumerate(code_free):
    if command == "[":
        stack.append(i)
    elif command == "]":
        start = stack.pop()
        bracket_map[start] = i
        bracket_map[i] = start

# Initializes the instruction pointer
instruction_pointer = 0

while instruction_pointer < len(code_free):
    i = code_free[instruction_pointer]  # Gets the current command

    if i == ">":  # Moves the pointer to the right (+1) or wraps to 0 if at the end
        cur_pointer = (cur_pointer + 1) % len(memory)
    elif i == "<":  # Moves the pointer to the left (-1) or wraps to the end if at 0
        cur_pointer = (cur_pointer - 1) % len(memory)
    elif i == "+":  # Increases the current memory cell by 1
        memory[cur_pointer] = (memory[cur_pointer] + 1) % 256
    elif i == "-":  # Decreases the current memory cell by 1
        memory[cur_pointer] = (memory[cur_pointer] - 1) % 256
    elif i == ".":  # Outputs the current memory cell as a character
        print(chr(memory[cur_pointer]), end="")
    elif i == ",":  # Inputs a character or integer into the current memory cell
        user_input = input("in: ")
        if len(user_input) == 1:
            memory[cur_pointer] = ord(user_input)
        else:
            try:
                value = int(user_input)
                if 0 <= value <= 255:
                    memory[cur_pointer] = value
                else:
                    print("Error: Input must be a single character or an integer between 0 and 255")
            except ValueError:
                print("Error: Invalid input. Input must be a single character or an integer between 0 and 255")
    elif i == "[":  # Detects the start of a loop
        if memory[cur_pointer] == 0:
            instruction_pointer = bracket_map[instruction_pointer]
    elif i == "]":  # Detects the end of a loop
        if memory[cur_pointer] != 0:
            instruction_pointer = bracket_map[instruction_pointer]
    instruction_pointer += 1  # Move to the next command

print("")  # Print a newline at the end, for readability in the terminal
# TODO: add comments to show in the terminal

File: Kotlin Compiler/compiler.kt
//TODO: add error handling

import java.nio.file.Paths
import java.nio.file.Files

fun executeCode(code: String, memory: ByteArray) {
    var curPointer: Int = 0
    var instructionPointer = 0

    // Preprocess loop jumps
    val jumpTable = mutableMapOf<Int, Int>()
    val openBracketStack = mutableListOf<Int>()
    for (i in code.indices) {
        when (code[i]) {
            '[' -> openBracketStack.add(i)
            ']' -> {
                val openIndex = openBracketStack.removeAt(openBracketStack.size - 1)
                jumpTable[openIndex] = i
                jumpTable[i] = openIndex
            }
        }
    }

    while (instructionPointer < code.length) {
        when (code[instructionPointer]) {
            '>' -> curPointer = (curPointer + 1) % memory.size // moves the pointer to the right, or wraps around to 0
            '<' -> curPointer = (curPointer - 1 + memory.size) % memory.size // moves the pointer to the left, or wraps around to the end
            '+' -> memory[curPointer] = (memory[curPointer] + 1).toByte() // increments the value at the current pointer by 1
            '-' -> memory[curPointer] = (memory[curPointer] - 1).toByte() // decrements the value at the current pointer by 1
            '.' -> print(memory[curPointer].toInt().toChar())
            ',' -> {
                print("Input: ")
                val input = readLine() //TODO: Fix not getting input
                if (!input.isNullOrEmpty()) {
                    memory[curPointer] = input[0].code.toByte()
                }
            }
            '[' -> if (memory[curPointer] == 0.toByte()) {
                instructionPointer = jumpTable[instructionPointer] ?: instructionPointer // goes to the next instruction if the current value is 0
            }
            ']' -> if (memory[curPointer] != 0.toByte()) {
                instructionPointer = jumpTable[instructionPointer] ?: instructionPointer // goes to the next instruction if the current value is not 0
            }
        }
        instructionPointer++
    }
}

fun main() {
    val code = Files.readString(Paths.get("../target.bf")) // reads target file
    val validChars = setOf('+', '-', '>', '<', '.', ',', '[', ']') // sets characters that can be interpreted, otherwise throws the characters away as comments 
    // TODO: add comments to show in the terminal
    val codeFree = code.filter { it in validChars }

    println("$codeFree\n")

    val memory = ByteArray(30000) // can be changed if more or less memory is desired

    executeCode(codeFree, memory) // run the compiler

    println("\n") // newline at the end for readability
}
